util/gpool/spmc/spmcpool_test.go

```diff
@@ -0,0 +1,242 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spmc
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+
+	"github.com/pingcap/tidb/resourcemanager/pooltask"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/gpool"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPool(t *testing.T) {
+	type ConstArgs struct {
+		a int
+	}
+	myArgs := ConstArgs{a: 10}
+	// init the pool
+	// input type， output type, constArgs type
+	pool, err := NewSPMCPool[int, int, ConstArgs, any, pooltask.NilContext]("TestPool", 10)
+	require.NoError(t, err)
+	pool.SetConsumerFunc(func(task int, constArgs ConstArgs, ctx any) int {
+		return task + constArgs.a
+	})
+	taskCh := make(chan int, 10)
+	for i := 1; i < 11; i++ {
+		taskCh <- i
+	}
+	pfunc := func() (int, error) {
+		select {
+		case task := <-taskCh:
+			return task, nil
+		default:
+			return 0, gpool.ErrProducerClosed
+		}
+	}
+	// add new task
+	resultCh, control := pool.AddProducer(pfunc, myArgs, pooltask.NilContext{}, WithConcurrency(4))
+
+	var count atomic.Uint32
+	var wg sync.WaitGroup
+	wg.Add(1)
+	exitCh2 := make(chan struct{})
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case result := <-resultCh:
+				count.Add(1)
+				require.Greater(t, result, 10)
+			case <-exitCh2:
+				return
+			}
+		}
+	}()
+	// Waiting task finishing
+	control.Wait()
+	close(exitCh2)
+	wg.Wait()
+	require.Equal(t, uint32(10), count.Load())
+
+	// close pool
+	pool.ReleaseAndWait()
+}
+
+func TestPoolWithEnoughCapa(t *testing.T) {
+	const RunTimes = 1000
+	p, err := NewSPMCPool[struct{}, struct{}, int, any, pooltask.NilContext]("TestPoolWithEnoughCapa", 30, WithExpiryDuration(DefaultExpiredTime))
+	require.NoError(t, err)
+	defer p.ReleaseAndWait()
+	p.SetConsumerFunc(func(a struct{}, b int, c any) struct{} {
+		return struct{}{}
+	})
+	var twg util.WaitGroupWrapper
+	for i := 0; i < 3; i++ {
+		twg.Run(func() {
+			sema := make(chan struct{}, 10)
+			var wg util.WaitGroupWrapper
+			exitCh := make(chan struct{})
+			wg.Run(func() {
+				for j := 0; j < RunTimes; j++ {
+					sema <- struct{}{}
+				}
+				close(exitCh)
+			})
+			producerFunc := func() (struct{}, error) {
+				for {
+					select {
+					case <-sema:
+						return struct{}{}, nil
+					default:
+						select {
+						case <-exitCh:
+							return struct{}{}, gpool.ErrProducerClosed
+						default:
+						}
+					}
+				}
+			}
+			resultCh, ctl := p.AddProducer(producerFunc, RunTimes, pooltask.NilContext{}, WithConcurrency(6))
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					select {
+					case <-resultCh:
+					default:
+						if ctl.IsProduceClose() {
+							return
+						}
+					}
+				}
+			}()
+			ctl.Wait()
+			wg.Wait()
+		})
+	}
+	twg.Wait()
+}
+
+func TestPoolWithoutEnoughCapa(t *testing.T) {
```

Comment:
The difference between this test and the test of `TestPoolWithEnoughCapa` above, doesn't seem obvious, it seems the same, right?

Reply:
I have reset the parameter between the two test cases.